File: orb/ios/Classes/Orb.swift
import Flutter
import os.log


@available(iOS 10.0, *)
extension OSLog {
    private static var subsystem = Bundle.main.bundleIdentifier!
    
    static let orb = OSLog(subsystem: subsystem, category: "Orb")
}


@objc public class Orb: NSObject {
    public let engine: FlutterEngine
    public let deviceId: String?
    public var deviceToken: String?
    public var ready: Bool {
        get { return _ready }
    }
    private var _ready: Bool = false
    private var _onReady: (() -> Void)? = nil
    
    public init(engine: FlutterEngine? = nil, name: String = "meya") {
        self.engine = engine ?? FlutterEngine(name: name)
        self.deviceId = UIDevice.current.identifierForVendor?.uuidString
    }
    
    deinit {
        Orb.log("Orb is being deinitialized")
    }
    
    public func initCallbacks() {
        if let plugin = Orb.getPlugin(self.engine) {
            plugin.ready = { [unowned self] in
                self._ready = true
                if let onReady = self._onReady {
                    onReady()
                }
            }
        } else {
            Orb.log("Orb plugin is not registered with the Flutter engine. Make sure you initialize Orb correctly")
        }
    }
    
    public func connect(options: OrbConnectionOptions, result: FlutterResult? = nil) {
        options.deviceId = options.deviceId ?? self.deviceId
        options.deviceToken = options.deviceToken ?? self.deviceToken
        Orb.getPlugin(self.engine)?.connect(options: options, result: result)
    }
    
    public func disconnect(logOut: Bool = false, result: FlutterResult? = nil) {
        Orb.getPlugin(self.engine)?.disconnect(logOut: logOut, result: result)
    }
    
    public func publishEvent(
        event: Dictionary<String, Any?>,
        result: FlutterResult? = nil
    ) {
        Orb.getPlugin(self.engine)?.publishEvent(event: event, result: result)
    }
    
    public func onReady(_ callback: @escaping () -> Void) {
        if (ready) {
            Orb.log("Orb is already running, this callback will not be called until the provided engine is restarted")
        }
        _onReady = callback
    }
    
    public func onConnnected(_ callback: @escaping () -> Void) {
        Orb.getPlugin(engine)?.connected = callback
    }
    
    public func onDisconnected(_ callback: @escaping () -> Void) {
        Orb.getPlugin(engine)?.disconnected = callback
    }
    
    public func onFirstConnect(
        _ callback: @escaping ([Dictionary<String, Any?>]) -> Void,
        result: FlutterResult? = nil
    ) {
        if let plugin = Orb.getPlugin(self.engine) {
            plugin.subscribe(name: "firstConnect", result: result)
            plugin.firstConnect = callback
        }
    }
    
    public func onReconnect(
        _ callback: @escaping ([Dictionary<String, Any?>]) -> Void,
        result: FlutterResult? = nil
    ) {
        if let plugin = Orb.getPlugin(self.engine) {
            plugin.subscribe(name: "reconnect", result: result)
            plugin.reconnect = callback
        }
    }
    
    public func onEvent(
        _ callback: @escaping (Dictionary<String, Any>, [Dictionary<String, Any?>]) -> Void,
        result: FlutterResult? = nil
    ) {
        if let plugin = Orb.getPlugin(self.engine) {
            plugin.subscribe(name: "event", result: result)
            plugin.event = callback
        }
    }
    
    public func onEventStream(
        _ callback: @escaping ([Dictionary<String, Any?>]) -> Void,
        result: FlutterResult? = nil
    ) {
        if let plugin = Orb.getPlugin(self.engine) {
            plugin.subscribe(name: "eventStream", result: result)
            plugin.eventStream = callback
        }
    }
    
    public func onCloseUi(_ callback: @escaping () -> Void) {
        Orb.getPlugin(engine)?.closeUi = callback
    }
    
    public func viewController(nibName: String? = nil, bundle: Bundle? = nil) -> FlutterViewController {
        return FlutterViewController(engine: engine, nibName: nibName, bundle: bundle)
    }
    
    static func getPlugin(_ engine: FlutterEngine) -> SwiftOrbPlugin? {
        return engine.valuePublished(byPlugin: "OrbPlugin") as? SwiftOrbPlugin
    }
    
    static func log(_ message: StaticString, _ args: CVarArg...) {
        if #available(iOS 10.0, *) {
            os_log(message, log: OSLog.orb, args)
        } else {
            print(message)
        }
    }
}

File: orb/android/src/main/java/ai/meya/orb/Orb.java
package ai.meya.orb;

import android.annotation.SuppressLint;
import android.content.Context;
import android.provider.Settings;
import android.util.Log;

import java.util.List;
import java.util.Map;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import io.flutter.embedding.engine.FlutterEngine;
import io.flutter.embedding.engine.dart.DartExecutor;
import io.flutter.embedding.engine.plugins.PluginRegistry;


public class Orb {
    private static final String TAG = "Orb";

    public String deviceId;
    public String deviceToken;
    public boolean ready = false;

    private ReadyListener _onReady;

    public interface ReadyListener {
        void onReady();
    }

    public interface ConnectedListener {
        void onConnected();
    }

    public interface DisconnectedListener {
        void onDisconnected();
    }

    public interface  FirstConnectListener {
        void onFirstConnect(List<Map<String, Object>> eventStream);
    }

    public interface  ReconnectListener {
        void onReconnect(List<Map<String, Object>> eventStream);
    }

    public interface  EventListener {
        void onEvent(Map<String, Object> event, List<Map<String, Object>> eventStream);
    }

    public interface  EventStreamListener {
        void onEventStream(List<Map<String, Object>> eventStream);
    }

    public interface CloseUiListener {
        void onCloseUi();
    }

    public FlutterEngine engine;

    @SuppressLint("HardwareIds")
    public Orb(@NonNull Context context, @NonNull FlutterEngine engine) {
        this.engine = engine;
        this.deviceId = Settings.Secure.getString(
                context.getContentResolver(), Settings.Secure.ANDROID_ID
        );
        OrbPlugin plugin = getPlugin(engine);
        if (plugin != null) {
            plugin.readyListener = new ReadyListener() {
                @Override
                public void onReady() {
                    Log.d(TAG, "Orb runtime ready");
                    ready = true;
                    if (_onReady != null) _onReady.onReady();
                }
            };
        } else {
            Log.e(
                    TAG,
                    "The Orb plugin has not been registered with the provided Flutter engine. Please ensure you've initialized Orb correctly."
            );
        }
    }

    public Orb(@NonNull Context context) {
        this(context, new FlutterEngine(context));
    }

    public void initialize() {
        engine.getDartExecutor().executeDartEntrypoint(
                DartExecutor.DartEntrypoint.createDefault()
        );
    }

    public void connect(OrbConnectionOptions options) {
        if (options.deviceId == null) options.deviceId = deviceId;
        if (options.deviceToken == null) options.deviceToken = deviceToken;
        OrbPlugin plugin = getPlugin(engine);
        if (plugin != null) plugin.connect(options);
    }

    public void disconnect() {
        OrbPlugin plugin = getPlugin(engine);
        if (plugin != null) plugin.disconnect(false);
    }
    
    public void disconnect(boolean logOut) {
        OrbPlugin plugin = getPlugin(engine);
        if (plugin != null) plugin.disconnect(logOut);
    }

    public void publishEvent(Map<String, Object> event) {
        OrbPlugin plugin = getPlugin(engine);
        if (plugin != null) plugin.publishEvent(event);
    }

    public void setOnReadyListener(Orb.ReadyListener listener) {
        if (ready) Log.d(TAG, "Orb is already running, this listener will not be called until the provided engine is restarted.");
        _onReady = listener;
    }

    public void setOnConnectedListener(Orb.ConnectedListener listener) {
        OrbPlugin plugin = getPlugin(engine);
        if (plugin == null) return;
        plugin.connectedListener = listener;
    }

    public void setOnDisconnectedListener(Orb.DisconnectedListener listener) {
        OrbPlugin plugin = getPlugin(engine);
        if (plugin == null) return;
        plugin.disconnectedListener = listener;
    }

    public void setOnFirstConnectListener(Orb.FirstConnectListener listener) {
        OrbPlugin plugin = getPlugin(engine);
        if (plugin == null) return;
        plugin.subscribe("firstConnect");
        plugin.firstConnectListener = listener;
    }

    public void setOnReconnectListener(Orb.ReconnectListener listener) {
        OrbPlugin plugin = getPlugin(engine);
        if (plugin == null) return;
        plugin.subscribe("reconnect");
        plugin.reconnectListener = listener;
    }

    public void setOnEventListener(Orb.EventListener listener) {
        OrbPlugin plugin = getPlugin(engine);
        if (plugin == null) return;
        plugin.subscribe("event");
        plugin.eventListener = listener;
    }

    public void setOnEventStreamListener(Orb.EventStreamListener listener) {
        OrbPlugin plugin = getPlugin(engine);
        if (plugin == null) return;
        plugin.subscribe("eventStream");
        plugin.eventStreamListener = listener;
    }

    public void setOnCloseUiListener(Orb.CloseUiListener listener) {
        OrbPlugin plugin = getPlugin(engine);
        if (plugin == null) return;
        plugin.closeUiListener = listener;
    }

    @Nullable
    public static OrbPlugin getPlugin(FlutterEngine engine) {
        if (!checkEngine(engine)) return null;

        PluginRegistry registry = engine.getPlugins();
        OrbPlugin plugin = (OrbPlugin) registry.get(OrbPlugin.class);

        if (plugin == null) {
            Log.e(TAG, "Could not get OrbPlugin.");
            return null;
        } else {
            return plugin;
        }
    }

    private static boolean checkEngine(FlutterEngine engine) {
        if (engine == null) {
            Log.e(TAG, "Orb not initialized yet.");
            return false;
        } else {
            return true;
        }
    }
}
